tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransaction.java

```diff
@@ -0,0 +1,173 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.tinkergraph.structure;
+
+import org.apache.tinkerpop.gremlin.process.traversal.TraversalSource;
+import org.apache.tinkerpop.gremlin.structure.Transaction;
+import org.apache.tinkerpop.gremlin.structure.util.AbstractThreadLocalTransaction;
+import org.apache.tinkerpop.gremlin.structure.util.TransactionException;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+final class TinkerTransaction extends AbstractThreadLocalTransaction {
+
+    private static final String TX_CONFLICT = "Conflict: element modified in another transaction";
+
+    private static final long NOT_STARTED = -1;
+
+    private static final AtomicLong openedTx;
+    private final ThreadLocal<Long> txNumber = ThreadLocal.withInitial(() -> NOT_STARTED);
+    private final ThreadLocal<Set<TinkerElementContainer<TinkerVertex>>> txChangedVertices = new ThreadLocal<>();
+    private final ThreadLocal<Set<TinkerElementContainer<TinkerEdge>>> txChangedEdges = new ThreadLocal<>();
+
+    private final TinkerTransactionGraph graph;
+
+    static {
+        openedTx = new AtomicLong(0);
+    }
+
+    public TinkerTransaction(final TinkerTransactionGraph g) {
+        super(g);
+        graph = g;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return txNumber.get() != NOT_STARTED;
+    }
+
+    @Override
+    public <T extends TraversalSource> T begin() {
+        doOpen();
+        return super.begin();
+    }
+
+    @Override
+    protected void doOpen() {
+        if (isOpen())
+            Transaction.Exceptions.transactionAlreadyOpen();
+
+        txNumber.set(openedTx.getAndIncrement());
+    }
+
+    protected long getTxNumber() {
+        if (!isOpen()) txNumber.set(openedTx.getAndIncrement());
+        return txNumber.get();
+    }
+
+    protected <T extends TinkerElement> void touch(TinkerElementContainer<T> container) {
+        if (!isOpen()) txNumber.set(openedTx.getAndIncrement());
+
+        T element = container.getUnmodified();
+        if (null == element) element = container.getModified();
+        if (element instanceof TinkerVertex) {
+            if (null == txChangedVertices.get())
+                txChangedVertices.set(new HashSet<>());
+            txChangedVertices.get().add((TinkerElementContainer<TinkerVertex>) container);
+        } else {
+            if (null == txChangedEdges.get())
+                txChangedEdges.set(new HashSet<>());
+            txChangedEdges.get().add((TinkerElementContainer<TinkerEdge>) container);
+        }
+    }
+
+    @Override
+    protected void doCommit() throws TransactionException {
```

Comment:
I think we need a narrative here (Javadoc) explaining the thought process and approach. It's a bit tough to decipher just looking at the code and it will be hard for others to work on this code later.